staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -398,37 +413,110 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
+	}
+	return obj, nil
+}
+
+func (t *tracker) create(gvr schema.GroupVersionResource, obj runtime.Object) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	objs, ok := t.objects[gvr]
+	if !ok {
+		objs = make(map[types.NamespacedName]runtime.Object)
+		t.objects[gvr] = objs
+	}
+
+	metaObj, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	namespacedName := types.NamespacedName{Namespace: metaObj.GetNamespace(), Name: metaObj.GetName()}
+	if _, ok = objs[namespacedName]; ok {
+		return errors.NewAlreadyExists(gvr.GroupResource(), metaObj.GetName())
+	}
+
+	objs[namespacedName] = obj
+
+	for _, w := range t.getWatches(gvr, metaObj.GetNamespace()) {
+		// To avoid the object from being accidentally modified by watcher
+		w.Add(obj.DeepCopyObject())
 	}
 
-	_, ok := t.objects[gvr]
+	return nil
+}
+
+func shouldDeleteDuringUpdate(oldObj, newObj runtime.Object) bool {
+	oldMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		panic(err)
+	}
+	newMeta, err := meta.Accessor(newObj)
+	if err != nil {
+		panic(err)
+	}
+	if len(newMeta.GetFinalizers()) > 0 {
+		// don't delete with finalizers remaining in the new object
+		return false
+	}
+	if oldMeta.GetDeletionTimestamp() == nil {
+		// don't delete if the existing object hasn't had a delete request made
+		return false
+	}
+	if ns, ok := newObj.(*v1.Namespace); ok && len(ns.Spec.Finalizers) > 0 {
+		// don't delete namespaces with finalizers remaining in the new object
+		return false
+	}
+	// delete if the existing object has no grace period or a grace period of 0
+	return oldMeta.GetDeletionGracePeriodSeconds() == nil || *oldMeta.GetDeletionGracePeriodSeconds() == 0
+
+}
+
+func (t *tracker) update(gvr schema.GroupVersionResource, obj runtime.Object) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	objs, ok := t.objects[gvr]
 	if !ok {
-		t.objects[gvr] = make(map[types.NamespacedName]runtime.Object)
+		return errors.NewNotFound(gvr.GroupResource(), "")
 	}
 
+	newMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
 	namespacedName := types.NamespacedName{Namespace: newMeta.GetNamespace(), Name: newMeta.GetName()}
-	if _, ok = t.objects[gvr][namespacedName]; ok {
-		if replaceExisting {
-			for _, w := range t.getWatches(gvr, ns) {
-				// To avoid the object from being accidentally modified by watcher
-				w.Modify(obj.DeepCopyObject())
-			}
-			t.objects[gvr][namespacedName] = obj
-			return nil
-		}
-		return errors.NewAlreadyExists(gr, newMeta.GetName())
+	oldObj, ok := objs[namespacedName]
+	if !ok {
+		return errors.NewNotFound(gvr.GroupResource(), newMeta.GetName())
 	}
 
-	if replaceExisting {
-		// Tried to update but no matching object was found.
-		return errors.NewNotFound(gr, newMeta.GetName())
+	oldMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		return err
+	}
+	if !oldMeta.GetDeletionTimestamp().Equal(newMeta.GetDeletionTimestamp()) {
+		return errors.NewBadRequest("metadata.deletionTimestamp is immutable")
+	}
```

Comment:
I don't see a good reason to reimplement this specific bit of server-side validation (this is just one of many things checked inside ValidateObjectMetaAccessorUpdate)

Also, there's server-side defaulting / overriding that happens for fields like uid / creationTimestamp / deletionTimestamp / deletionGracePeriodSeconds, so erroring here will not actually match server-side behavior:

```go
	// an update can never remove/change a deletion timestamp
	if !oldMeta.GetDeletionTimestamp().IsZero() {
		objectMeta.SetDeletionTimestamp(oldMeta.GetDeletionTimestamp())
	}
```